sky/packages/sky/lib/sky_tool

```diff
@@ -254,6 +284,355 @@ class StopSky(object):
+            return cls._has_ios_deploy
+        cmd = [
+            'which',
+            'ios-deploy'
+        ]
+        out = subprocess.check_output(cmd)
+        match = re.search(r'ios-deploy', out)
+        cls._has_ios_deploy = match is not None
+        return cls._has_ios_deploy
+
+    _is_connected = False
+    @classmethod
+    def is_connected(cls):
+        if not cls.has_ios_deploy():
+            return False
+        if cls._is_connected:
+            return True
+        cmd = [
+            'ios-deploy',
+            '--detect',
+            '--timeout',
+            '1'
+        ]
+        out = subprocess.check_output(cmd)
+        match = re.search(r'\[\.\.\.\.\] Found [^\)]*\) connected', out)
+        cls._is_connected = match is not None
+        return cls._is_connected
+
+    @classmethod
+    def install_app(cls, ios_app_path):
+        if not cls.has_ios_deploy():
+            return
+        cmd = [
+            'ios-deploy',
+            '--justlaunch',
+            '--timeout',
+            '10', # Smaller timeouts cause it to exit before having launched the app
+            '--bundle',
+            ios_app_path
+        ]
+        subprocess.check_call(cmd)
+
+    @classmethod
+    def copy_file(cls, bundle_id, local_path, device_path):
+        if not cls.has_ios_deploy():
+            return
+        cmd = [
+            'ios-deploy',
+            '-t',
+            '1',
+            '--bundle_id',
+            bundle_id,
+            '--upload',
+            local_path,
+            '--to',
+            device_path
+        ]
+        subprocess.check_call(cmd)
+
+
+class IOSSimulator(object):
+    @classmethod
+    def is_booted(cls):
+        return cls.get_simulator_device_id() is not None
+
+    _device_id = None
+    @classmethod
+    def get_simulator_device_id(cls):
+        if cls._device_id is not None:
+            return cls._device_id
+        cmd = [
+            'xcrun',
+            'simctl',
+            'list',
+            'devices',
+        ]
+        out = subprocess.check_output(cmd)
+        match = re.search(r'[^\(]+\(([^\)]+)\) \(Booted\)', out)
+        if match is not None and match.group(1) is not None:
+            cls._device_id = match.group(1)
+            return cls._device_id
+        else:
+            logging.warning('No running simulators found')
+            # TODO: Maybe start the simulator?
+            return None
+        if err is not None:
+            print(err)
+            exit(-1)
+
+    _simulator_path = None
+    @classmethod
+    def get_simulator_path(cls):
+        if cls._simulator_path is not None:
+            return cls._simulator_path
+        home_dir = os.path.expanduser("~")
+        device_id = cls.get_simulator_device_id()
+        if device_id is None:
+            # TODO: Maybe start the simulator?
+            return None
+        cls._simulator_path = home_dir + '/Library/Developer/CoreSimulator/Devices/' + device_id
```

Comment:
typically I use os.path.join instead of + since it handles foo/ + /bar, etc than just raw + does.